### ISSUE TITLE
How model see different lanes?

### ISSUE THREAD

**Issue by zguo008** [cardwing/Codes-for-Lane-Detection#89]
Hi,

I have a question about the model itself. From the lanenet model and encoder stage I found that the out logits is basically a smoothened binary prediction. However I wonder how the model is trained to see 4 lanes seperately?

I looked into details of the code, at lanenet merge model.py ,ret['instance seg loss']=binary_seg_loss.This confuses me a bit.

Could you please share some insight how the model see different lanes on its own?(It would be nice if you can tell me in which part of the code the model see different lanes)

Best regards,

ZX

**Comment by cardwing** [comment]
The LaneNet model first generates a binary prediction and then uses the clustering module to classify the lane pixels into different lane instances.

**Comment by zguo008** [comment]
@cardwing Hi, thank you for your reply. I used the weekend to check the code again in details. I think the cluster module u mean is at 'lanenet model, function: loss'. In other words, the model is trained to reduce the loss for 5 classes (background and 4 lanes). As a result the module can see 4 lanes and background separately.

I have some trouble understanding these few lines of code:
 binary_label_reshape = tf.one_hot(binary_label_reshape, depth=5)
 class_weights = tf.constant([[0.4, 1.0, 1.0, 1.0, 1.0]])
 weights_loss = tf.reduce_sum(tf.multiply(binary_label_reshape, class_weights), 2)
binary_segmentation_loss=tf.losses.softmax_cross_entropy(onehot_labels=binary_label_reshape,logits=decode_logits_reshape,weights=weights_loss)

My question is: 
1.binary_label picture is first reshape to one line of pixels and then one-hot encoded. How can a binary picture be encoded to 5 classes representing background and lanes?
2.When calculating softmax cross entropy loss, the decode logits is a binary prediction and binary_label_reshape is a instance prediction, how to reduce the loss like this? 

Thanks in advance

**Comment by SuperPengXue** [comment]
@zguo008   hi the code "class_weights = tf.constant([[0.4, 1.0, 1.0, 1.0, 1.0]])",I think it's to solve the problem of sample imbalance.

**Comment by cardwing** [comment]
@zguo008, sorry for the old misleading answer. LaneNet of this repo is just SCNN-Tensorflow and it outputs a multi-channel probability map and the lane existence prediction vector. There are only two losses, e.g., the instance segmentation loss and the lane existence prediction loss.

**Comment by zguo008** [comment]
Thank you all!
